File: benchmarks/families/transformer.py
#!/usr/bin/env python3
"""The transformer family in PyTorch, defined once, driven four ways.

PLAN.md; the counterpart of transformer.cpp. Encoder-decoder, d_model 512, 8
heads, feed-forward 2048, 6 layers -- the "Attention Is All You Need" base
model -- with scaled token embeddings, sinusoidal positions and a final
projection to the vocabulary, which is what transformer.cpp builds.

  transformer.py train    <corpus> <corpus> [epochs] [batch,...] [d_model] [layers] [dev] [prec]
  transformer.py infer    <corpus>          [reps]   [batch,...] [d_model] [layers] [dev] [prec]
  transformer.py capacity <corpus>          [batch]              [d_model] [layers] [dev] [prec]

**On tokenisation.** The two engines do not share a tokeniser: OpenNN's
LanguageDataset derives its own vocabulary and sequence length, and
reimplementing that here would be a second thing to keep in step. What must
match is the tensor *shape*, because that is what the arithmetic depends on --
sequence length, vocabulary size and batch. Both are printed by both engines
and recorded in the artifact, so a divergence shows up as data rather than
hiding inside a throughput number. The token identities themselves are
irrelevant to a speed measurement; their count is not.
"""

from __future__ import annotations

import collections
import contextlib
import math
import os
import sys
import time
from pathlib import Path

import torch
from torch import nn

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

SEED = 42
VOCAB_CAP = 20_000          # LanguageDataset's own cap, so both sides agree

class PositionalEncoding(nn.Module):
    def __init__(self, d_model: int, max_len: int):
        super().__init__()
        encoding = torch.zeros(max_len, d_model)
        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
        divisor = torch.exp(torch.arange(0, d_model, 2).float()
                            * (-math.log(10000.0) / d_model))
        encoding[:, 0::2] = torch.sin(position * divisor)
        encoding[:, 1::2] = torch.cos(position * divisor)
        self.register_buffer("encoding", encoding.unsqueeze(0))

    def forward(self, x):
        return x + self.encoding[:, : x.size(1)]

class Seq2Seq(nn.Module):
    def __init__(self, vocab: int, sequence: int, d_model: int, layers: int):
        super().__init__()
        heads = max(1, d_model // 64)
        self.scale = math.sqrt(d_model)
        self.source_embedding = nn.Embedding(vocab, d_model)
        self.target_embedding = nn.Embedding(vocab, d_model)
        self.positions = PositionalEncoding(d_model, sequence + 1)
        # Encoder and decoder built directly rather than through
        # nn.Transformer, which appends a final LayerNorm to each stack.
        # Those two norms are 2 * (512 weight + 512 bias) = 2,048 parameters
        # OpenNN's Transformer does not have, and two extra normalisations per
        # forward pass over every position. norm=None drops them, so the two
        # engines run the same graph rather than merely a similar one.
        encoder_layer = nn.TransformerEncoderLayer(d_model, heads, 4 * d_model,
                                                   batch_first=True)
        decoder_layer = nn.TransformerDecoderLayer(d_model, heads, 4 * d_model,
                                                   batch_first=True)
        self.encoder = nn.TransformerEncoder(encoder_layer, layers, norm=None)
        self.decoder = nn.TransformerDecoder(decoder_layer, layers, norm=None)
        self.output = nn.Linear(d_model, vocab)

    def forward(self, source, target):
        s = self.positions(self.source_embedding(source) * self.scale)
        t = self.positions(self.target_embedding(target) * self.scale)
        return self.output(self.decoder(t, self.encoder(s)))

def build(vocab: int, sequence: int, opts: dict) -> nn.Module:
    """The transformer family. Nothing else here constructs the network."""
    torch.manual_seed(SEED)
    return Seq2Seq(vocab, sequence, opts["d_model"], opts["layers"]).to(opts["device"])

def load_corpus(path: str, sequence: int | None = None) -> tuple[torch.Tensor, torch.Tensor, int, int]:
    """`source <TAB> target` to padded id tensors, and the shapes derived.

    Whitespace tokens, ranked by frequency and capped at VOCAB_CAP, with 0
    reserved for padding. Sequence length is the longest pair unless one is
    given, so the padded shape is a property of the corpus rather than a knob.
    """
    sources, targets = [], []
    counter: collections.Counter = collections.Counter()

    with open(path, encoding="utf-8") as handle:
        for line in handle:
            parts = line.rstrip("\n").split("\t")
            if len(parts) != 2:
                continue
            source, target = parts[0].split(), parts[1].split()
            counter.update(source)
            counter.update(target)
            sources.append(source)
            targets.append(target)

    # OpenNN's WordLevelTokenizer reserves [PAD]=0, [UNK]=1, [START]=2,
    # [END]=3, and brackets each sequence with START and END. Mirroring that
    # is not cosmetic: those two positions are two more rows of attention per
    # sequence, and without them the padded length differed by exactly 2.
    PAD, UNK, START, END = 0, 1, 2, 3
    vocabulary = {token: index + 4
                  for index, (token, _) in enumerate(counter.most_common(VOCAB_CAP - 4))}

    longest = max(max(len(s) for s in sources), max(len(t) for t in targets))
    length = sequence or longest + 2

    def encode(rows):
        tensor = torch.full((len(rows), length), PAD, dtype=torch.long)
        for row_index, row in enumerate(rows):
            ids = [START] + [vocabulary.get(t, UNK) for t in row[: length - 2]] + [END]
            tensor[row_index, : len(ids)] = torch.tensor(ids, dtype=torch.long)
        return tensor

    return encode(sources), encode(targets), VOCAB_CAP, length

def parse_opts(argv: list[str], first: int) -> dict:
    def at(index: int, default: str) -> str:
        return argv[index] if len(argv) > index else default

    device = at(first + 2, "cuda")
    precision = at(first + 3, "fp32")

    if device == "cuda" and not torch.cuda.is_available():
        raise SystemExit("cuda requested but not available")

    allow_tf32 = precision != "strict"
    torch.backends.cuda.matmul.allow_tf32 = allow_tf32
    torch.backends.cudnn.allow_tf32 = allow_tf32

    return {"d_model": int(at(first, "512")), "layers": int(at(first + 1, "6")),
            "device": device, "precision": precision,
            "autocast": precision == "bf16"}

def autocast_ctx(opts: dict):
    if not opts["autocast"]:
        return contextlib.nullcontext()
    return torch.autocast(device_type=opts["device"], dtype=torch.bfloat16)

def compiled(fn, opts: dict):
    mode = os.environ.get("PT_COMPILE_MODE", "default")
    if mode == "eager" or opts["device"] != "cuda":
        return fn, "eager"
    return torch.compile(fn, mode=None if mode == "default" else mode), f"compile:{mode}"

def batches_of(text: str) -> list[int]:
    return [int(part) for part in text.split(",") if part]

def sync(opts: dict) -> None:
    if opts["device"] == "cuda":
        torch.cuda.synchronize()

def report(batch: int, seconds: list[float], sequences: int, tokens_per_sequence: int) -> None:
    seconds = sorted(seconds)
    median = seconds[len(seconds) // 2]
    print(f"batch_{batch}_samples_per_sec={int(sequences / median)}"
          f" median_epoch_s={median:.6g}")
    print(f"batch_{batch}_tokens_per_sec={int(sequences * tokens_per_sequence / median)}")
    print(f"batch_{batch}_epoch_times={','.join(f'{t:.6g}' for t in seconds)}", flush=True)

def train_like(argv: list[str], mode: str) -> int:
    epochs = int(argv[4]) if len(argv) > 4 else 1
    batches = batches_of(argv[5] if len(argv) > 5 else "32")
    opts = parse_opts(argv, 6)

    print(f"engine=pytorch\nmode={mode}\ndevice={opts['device']}")
    source, target, vocab, sequence = load_corpus(argv[2])
    print(f"samples={source.shape[0]} sequence={sequence} "
          f"input_vocab={vocab} target_vocab={vocab} "
          f"d_model={opts['d_model']} layers={opts['layers']}")

    source = source.to(opts["device"])
    target = target.to(opts["device"])
    warmup = 1 if mode == "train" else 0

    for batch in batches:
        model = build(vocab, sequence, opts)
        if batch == batches[0]:
            print(f"parameters={sum(p.numel() for p in model.parameters())}", flush=True)
        loss_fn = nn.CrossEntropyLoss()
        optimizer = torch.optim.Adam(model.parameters(), lr=1.0e-4)
        ctx = autocast_ctx(opts)

        def step(s, t):
            optimizer.zero_grad(set_to_none=True)
            with ctx:
                logits = model(s, t)
                loss = loss_fn(logits.reshape(-1, vocab), t.reshape(-1))
            loss.backward()
            optimizer.step()

        step_fn, how = compiled(step, opts)
        starts = range(0, source.shape[0] - batch + 1, batch)

        def run_epoch():
            model.train()
            for start in starts:
                step_fn(source[start:start + batch], target[start:start + batch])
            sync(opts)

        for _ in range(warmup):
            run_epoch()

        print(f"TIMED_START_UNIX={time.time():.3f}", flush=True)
        times = []
        for _ in range(epochs):
            mark = time.perf_counter()
            run_epoch()
            times.append(time.perf_counter() - mark)
        print(f"TIMED_END_UNIX={time.time():.3f}", flush=True)

        print(f"batch_{batch}_mode={how}")
        report(batch, times, len(list(starts)) * batch, sequence)

    print("RESULT=OK")
    return 0

def infer(argv: list[str]) -> int:
    reps = int(argv[3]) if len(argv) > 3 else 1
    batches = batches_of(argv[4] if len(argv) > 4 else "32")
    opts = parse_opts(argv, 5)

    print(f"engine=pytorch\nmode=infer\ndevice={opts['device']}")
    source, target, vocab, sequence = load_corpus(argv[2])
    print(f"samples={source.shape[0]} sequence={sequence} "
          f"input_vocab={vocab} target_vocab={vocab}")

    source = source.to(opts["device"])
    target = target.to(opts["device"])

    for batch in batches:
        model = build(vocab, sequence, opts).eval()
        if batch == batches[0]:
            print(f"parameters={sum(p.numel() for p in model.parameters())}", flush=True)
        forward, _ = compiled(model, opts)
        processed = (source.shape[0] // batch) * batch

        s, t = source[:batch], target[:batch]

        def run_pass():
            with torch.no_grad(), autocast_ctx(opts):
                for _ in range(processed // batch):
                    forward(s, t)
            sync(opts)

        run_pass()

        print(f"TIMED_START_UNIX={time.time():.3f}", flush=True)
        times = []
        for _ in range(reps):
            mark = time.perf_counter()
            run_pass()
            times.append(time.perf_counter() - mark)
        print(f"TIMED_END_UNIX={time.time():.3f}", flush=True)

        report(batch, times, processed, sequence)

    print("RESULT=OK")
    return 0

def capacity(argv: list[str]) -> int:
    batch = int(argv[3]) if len(argv) > 3 else 32
    opts = parse_opts(argv, 4)

    print(f"engine=pytorch\nmode=capacity\ndevice={opts['device']}\nbatch={batch}")

    try:
        source, target, vocab, sequence = load_corpus(argv[2])
        s = source[:batch].to(opts["device"])
        t = target[:batch].to(opts["device"])

        model = build(vocab, sequence, opts)
        print(f"parameters={sum(p.numel() for p in model.parameters())}", flush=True)
        loss_fn = nn.CrossEntropyLoss()
        optimizer = torch.optim.Adam(model.parameters(), lr=1.0e-4)

        with autocast_ctx(opts):
            loss = loss_fn(model(s, t).reshape(-1, vocab), t.reshape(-1))
        loss.backward()
        optimizer.step()
        sync(opts)
    except torch.cuda.OutOfMemoryError as error:
        print(f"fits=0\nreason={error}\nRESULT=OOM", flush=True)
        return 1
    except RuntimeError as error:
        if "out of memory" not in str(error).lower():
            raise
        print(f"fits=0\nreason={error}\nRESULT=OOM", flush=True)
        return 1

    print("fits=1\nRESULT=OK", flush=True)
    return 0

def main() -> int:
    argv = sys.argv
    mode = argv[1] if len(argv) > 1 else ""

    if mode in ("train", "quality") and len(argv) > 3:
        return train_like(argv, mode)
    if mode == "infer" and len(argv) > 2:
        return infer(argv)
    if mode == "capacity" and len(argv) > 2:
        return capacity(argv)

    print(__doc__, file=sys.stderr)
    return 2

if __name__ == "__main__":
    raise SystemExit(main())

File: benchmarks/families/transformer.cpp
// The transformer family, defined once, driven four ways.
//
// PLAN.md. The "Attention Is All You Need" base model -- d_model 512, 8
// heads, feed-forward 2048, 6 layers -- on WMT14 English-German, which is the
// corpus that paper trained and reported on and therefore the citable one.
// Heads and feed-forward width follow d_model by the paper's own ratios
// (d_model/64 heads, 4*d_model feed-forward) rather than being separate
// knobs, so a change to d_model cannot silently produce a model the paper
// would not recognise.
//
//   transformer train    <corpus> <corpus> [epochs] [batch,...] [d_model] [layers] [dev] [prec]
//   transformer infer    <corpus>          [reps]   [batch,...] [d_model] [layers] [dev] [prec]
//   transformer capacity <corpus>          [batch]              [d_model] [layers] [dev] [prec]
//
// Throughput is reported both per sequence and per token. Per token is what
// the literature quotes; per sequence is what the runner compares, and the
// two differ by the padded sequence length, which the corpus fixes.

#include <algorithm>
#include <chrono>
#include <filesystem>
#include <iomanip>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#ifdef __linux__
#include <unistd.h>
#endif

#include "opennn/core/configuration.h"
#include "opennn/core/tensor_operations.h"
#include "opennn/core/random_utilities.h"
#include "opennn/core/device_backend.h"
#include "opennn/core/tensor_types.h"
#include "opennn/dataset/language_dataset.h"
#include "opennn/neural_network/forward_propagation.h"
#include "opennn/models/models.h"
#include "opennn/training_strategy/adaptive_moment_estimation.h"
#include "opennn/training_strategy/training_strategy.h"

using namespace opennn;
using clock_type = chrono::steady_clock;

namespace
{

double resident_mb()
{
#ifdef __linux__
    // statm reports pages; the second field is resident.
    ifstream statm("/proc/self/statm");
    long total = 0, resident = 0;
    if (statm >> total >> resident)
        return double(resident) * double(sysconf(_SC_PAGESIZE)) / (1024.0 * 1024.0);
#endif
    return 0.0;
}

constexpr Index SEED = 42;
constexpr Index BF16_SDPA_MIN_SEQUENCE = 128;

struct Options
{
    Index d_model = 512;
    Index layers = 6;
    Device device = Device::CUDA;
    Type precision = Type::FP32;

    Index heads() const { return max(Index(1), d_model / 64); }
    Index feed_forward() const { return 4 * d_model; }
};

bool use_bf16_sdpa(const Options& options)
{
#if defined(OPENNN_HAS_CUDA) && CUDNN_VERSION >= 92500
    return options.device == Device::CUDA && options.precision == Type::BF16;
#else
    (void)options;
    return false;
#endif
}

unique_ptr<Transformer> build(LanguageDataset& dataset, const Options& options)
{
    set_seed(SEED);

    auto transformer = make_unique<Transformer>(
        dataset.get_shape("Input")[0],
        dataset.get_shape("Decoder")[0],
        dataset.get_input_vocabulary_size(),
        dataset.get_target_vocabulary_size(),
        options.d_model,
        options.heads(),
        options.feed_forward(),
        options.layers);

    // WMT14 rows are 130 tokens after START/END are added, so the library-wide
    // 192-token crossover puts them on the materialized path, which keeps a
    // full attention matrix per encoder/decoder attention layer. PyTorch uses
    // fused scaled-dot-product attention at this length, so leaving OpenNN
    // materialized compares one engine's guarded path against the other's fast
    // one. Both measured cells here run many iterations over the corpus, which
    // is the regime the fused path is for.
    //
    // Measured on this suite, sweeping sequence length with everything else
    // fixed, fused throughput beat materialized at every length tried, by
    // roughly 6% at 32 tokens rising to about 30% by 192 and 256. Those sweep
    // points are one launch each and the fused path varies by about 6% between
    // launches while the materialized one holds to 0.5%, so read them as a
    // trend rather than as figures. Five launches each way at 128 tokens give
    // medians of 4674 against 3643 samples/s, a 28% gain whose distributions do
    // not overlap at all -- 4231 slowest fused against 3653 fastest
    // materialized. The library default
    // stays at 192 regardless, and deliberately: the same sweep run as a single
    // pass reverses the result, because cuDNN plan construction costs 0.3-2.0 s
    // with nothing to amortize it against, leaving fused 1.6x slower at 256 and
    // 5x slower at 32. Sequence length is only a proxy for the thing that
    // actually decides this, which is how often the plan gets reused; a
    // benchmark cell reusing it across every batch of 4,096 samples sits firmly
    // on the fused side of that, and a caller doing one forward pass does not.
    //
    // The reference cuDNN 9.25 supports this graph; older runtimes stay on the
    // materialized path because some reject the 130-token plan outright.
    if(use_bf16_sdpa(options))
    {
        transformer->set_attention_sdpa_min_sequence_length(
            BF16_SDPA_MIN_SEQUENCE);
    }

    return transformer;
}

vector<Index> parse_batches(const string& text)
{
    vector<Index> batches;
    stringstream stream(text);

    for (string item; getline(stream, item, ',');)
        if (!item.empty())
            batches.push_back(stoll(item));

    return batches;
}

Options parse_options(int argc, char* argv[], int first)
{
    Options options;

    if (argc > first)     options.d_model = stoll(argv[first]);
    if (argc > first + 1) options.layers = stoll(argv[first + 1]);
    if (argc > first + 2) options.device = string(argv[first + 2]) == "cpu" ? Device::CPU : Device::CUDA;
    if (argc > first + 3) options.precision = string(argv[first + 3]) == "bf16" ? Type::BF16 : Type::FP32;

    return options;
}

AdaptiveMomentEstimation* configure(TrainingStrategy& strategy, Index batch)
{
    strategy.set_loss("CrossEntropyError3d");
    strategy.set_optimization_algorithm("AdaptiveMomentEstimation");

    auto* adam = dynamic_cast<AdaptiveMomentEstimation*>(strategy.get_optimization_algorithm());
    adam->set_batch_size(batch);
    adam->set_display(false);
    adam->set_display_period(1000000);
    adam->set_learning_rate(0.0001f);
    adam->set_joint_gradient_arena(true);

    return adam;
}

int usage()
{
    cerr << "usage: transformer train    <corpus> <corpus> [epochs] [batch,...] [d_model] [layers] [dev] [prec]\n"
            "       transformer infer    <corpus>          [reps]   [batch,...] [d_model] [layers] [dev] [prec]\n"
            "       transformer capacity <corpus>          [batch]              [d_model] [layers] [dev] [prec]\n";
    return 2;
}

void report(Index batch, const vector<double>& seconds, Index sequences, Index tokens_per_sequence)
{
    vector<double> sorted = seconds;
    sort(sorted.begin(), sorted.end());
    const double median = sorted[sorted.size() / 2];

    cout << "batch_" << batch << "_samples_per_sec=" << long(double(sequences) / median)
         << " median_epoch_s=" << median << "\n"
         << "batch_" << batch << "_tokens_per_sec="
         << long(double(sequences) * double(tokens_per_sequence) / median) << "\n"
         << "batch_" << batch << "_epoch_times=";
    for (size_t i = 0; i < seconds.size(); ++i) cout << (i ? "," : "") << seconds[i];
    cout << "\n" << flush;
}

}   // namespace

int main(int argc, char* argv[])
{
    // Each engine at its best, as PROTOCOL.md requires. The library defaults to
    // Eigen so a plain build behaves like a plain build; a build that has the
    // MKL kernels is told to use them here rather than inheriting them.
    Configuration::instance().set_blas(Blas::Mkl);
    cout << "blas=" << (blas_mkl_available() ? "mkl" : "eigen") << "\n";

    const string mode = argc > 1 ? argv[1] : "";

    if (mode == "train" || mode == "quality")
    {
        if (argc < 4) return usage();

        const Index epochs = argc > 4 ? stoll(argv[4]) : 1;
        const vector<Index> batches = parse_batches(argc > 5 ? argv[5] : "32");
        const Options options = parse_options(argc, argv, 6);

        Configuration::instance().set(options.device, options.precision);
        cout << "baseline_rss_mib=" << resident_mb() << "\n";
        cout << "engine=opennn\nmode=" << mode
             << "\ndevice=" << (options.device == Device::CPU ? "cpu" : "cuda") << "\n";

        cout << "dataset_opened=" << filesystem::absolute(argv[2]).string() << "\n" << flush;
        LanguageDataset dataset(argv[2]);

        dataset.set_sample_roles("Training");

        const Index samples = dataset.get_samples_number();
        const Index sequence = dataset.get_shape("Input")[0];

        cout << "samples=" << samples << " sequence=" << sequence
             << " input_vocab=" << dataset.get_input_vocabulary_size()
             << " target_vocab=" << dataset.get_target_vocabulary_size()
             << " d_model=" << options.d_model << " heads=" << options.heads()
             << " ff=" << options.feed_forward() << " layers=" << options.layers << "\n";

        const bool timing = mode == "train";
        const Index warmup = timing ? 1 : 0;

        for (const Index batch : batches)
        {
            auto network = build(dataset, options);
            if (batch == batches.front())
            {
                cout << "parameters=" << network->get_parameters_number() << "\n" << flush;
                if (use_bf16_sdpa(options))
                {
                    cout << "sdpa_min_sequence_length="
                         << BF16_SDPA_MIN_SEQUENCE << "\n" << flush;
                }
            }

            const bool graph = options.device == Device::CUDA
                               && getenv("OPENNN_NO_CUDA_GRAPH") == nullptr;

            TrainingStrategy strategy(network.get(), &dataset);
            auto* adam = configure(strategy, batch);
            adam->set_cuda_graph(graph);
            adam->set_maximum_epochs(warmup + epochs);

            const auto unix_now = []
            {
                return chrono::duration<double>(
                    chrono::system_clock::now().time_since_epoch()).count();
            };

            vector<double> epoch_seconds;
            auto previous_mark = clock_type::now();

            adam->post_epoch_callback = [&](Index epoch, float, float, NeuralNetwork*)
            {
                const auto now = clock_type::now();
                const double elapsed = chrono::duration<double>(now - previous_mark).count();
                previous_mark = now;

                if (epoch == warmup - 1)
                    cout << "TIMED_START_UNIX=" << fixed << setprecision(3)
                         << unix_now() << "\n" << defaultfloat;
                else if (epoch >= warmup)
                    epoch_seconds.push_back(elapsed);

                if (epoch == warmup + epochs - 1)
                    cout << "TIMED_END_UNIX=" << fixed << setprecision(3)
                         << unix_now() << "\n" << defaultfloat;
            };

            strategy.train();

            if (Index(epoch_seconds.size()) != epochs)
            {
                cerr << "epoch timing marks missing\n";
                return 1;
            }

            cout << "batch_" << batch << "_cuda_graph="
                 << (!graph ? "off"
                     : adam->get_cuda_graph_capture_failed() ? "failed" : "captured") << "\n";

            report(batch, epoch_seconds, (samples / batch) * batch, sequence);
        }
    }
    else if (mode == "infer")
    {
        if (argc < 3) return usage();

        const Index reps = argc > 3 ? stoll(argv[3]) : 1;
        const vector<Index> batches = parse_batches(argc > 4 ? argv[4] : "32");
        const Options options = parse_options(argc, argv, 5);

        Configuration::instance().set(options.device, options.precision);
        cout << "baseline_rss_mib=" << resident_mb() << "\n";
        cout << "engine=opennn\nmode=infer\ndevice="
             << (options.device == Device::CPU ? "cpu" : "cuda") << "\n";

        LanguageDataset dataset(argv[2]);
        dataset.set_sample_roles("Testing");

        const Index samples = dataset.get_samples_number();
        const Index sequence = dataset.get_shape("Input")[0];

        cout << "samples=" << samples << " sequence=" << sequence
             << " input_vocab=" << dataset.get_input_vocabulary_size()
             << " target_vocab=" << dataset.get_target_vocabulary_size() << "\n";

        for (const Index batch : batches)
        {
            auto network = build(dataset, options);
            if (batch == batches.front())
                cout << "parameters=" << network->get_parameters_number() << "\n" << flush;
            const Index processed = (samples / batch) * batch;

            ForwardPropagation forward_propagation(batch, network.get(),
                                                   ForwardPropagationMode::Inference);
            const bool graph = options.device == Device::CUDA
                               && getenv("OPENNN_NO_CUDA_GRAPH") == nullptr;
            forward_propagation.set_cuda_graph(graph);

            vector<Index> indices(size_t(batch), Index(0));
            for (Index k = 0; k < batch; ++k) indices[size_t(k)] = k;

            Batch data(batch, &dataset, network->get_config());
            data.fill(indices, dataset.get_feature_selection(), FillMode::Inference);

            const vector<TensorView>& inputs = data.get_inputs();

            const auto run_pass = [&]
            {
                for (Index i = 0; i + batch <= samples; i += batch)
                    network->calculate_outputs_resident(inputs, forward_propagation, false);

                // The clock stops when the work is done, not when it is
                // queued, matching torch.cuda.synchronize() in the PyTorch
                // driver. At this pass length the queue saturates and the
                // host blocks anyway, so this is worth about one graph
                // launch -- but a pass short enough to fit the async queue
                // would otherwise report launch throughput, which is how the
                // dense inference cell read 72.5M samples/s.
                if (options.device == Device::CUDA)
                    device::synchronize(device::get_compute_stream());
            };

            // Uploading parameters is a device operation -- a CPU-compiled
            // network has no device copy and `copy_parameters_device` throws,
            // which aborted every CPU inference cell in this family. The
            // warm-up pass still has to happen, so it is the upload that is
            // conditional, not the pass.
            network->calculate_outputs_resident(inputs, forward_propagation,
                                                options.device == Device::CUDA);
            run_pass();

            const auto unix_now = []
            {
                return chrono::duration<double>(
                    chrono::system_clock::now().time_since_epoch()).count();
            };

            cout << "TIMED_START_UNIX=" << fixed << setprecision(3)
                 << unix_now() << "\n" << defaultfloat << flush;

            vector<double> times;
            for (Index r = 0; r < reps; ++r)
            {
                const auto t0 = clock_type::now();
                run_pass();
                times.push_back(chrono::duration<double>(clock_type::now() - t0).count());
            }

            cout << "TIMED_END_UNIX=" << fixed << setprecision(3)
                 << unix_now() << "\n" << defaultfloat << flush;

            report(batch, times, processed, sequence);
            cout << "batch_" << batch << "_cuda_graph="
                 << (!graph ? "off"
                     : forward_propagation.cuda_graph_failed ? "failed"
                     : forward_propagation.inference_graph_exec ? "captured" : "warming")
                 << "\n" << flush;
        }
    }
    else if (mode == "capacity")
    {
        if (argc < 3) return usage();

        const Index batch = argc > 3 ? stoll(argv[3]) : 32;
        const Options options = parse_options(argc, argv, 4);

        Configuration::instance().set(options.device, options.precision);
        cout << "baseline_rss_mib=" << resident_mb() << "\n";
        cout << "engine=opennn\nmode=capacity\ndevice="
             << (options.device == Device::CPU ? "cpu" : "cuda")
             << "\nbatch=" << batch << "\n";

        try
        {
            LanguageDataset dataset(argv[2]);
            dataset.set_sample_roles("Training");

            auto network = build(dataset, options);
            cout << "parameters=" << network->get_parameters_number() << "\n" << flush;

            TrainingStrategy strategy(network.get(), &dataset);
            configure(strategy, batch)->set_maximum_epochs(1);
            strategy.train();
        }
        catch (const exception& error)
        {
            cout << "fits=0\nreason=" << error.what() << "\nRESULT=OOM\n" << flush;
            return 1;
        }

        cout << "fits=1\nRESULT=OK\n" << flush;
        return 0;
    }
    else
    {
        return usage();
    }

    cout << "RESULT=OK\n";

    return 0;
}
